### ISSUE TITLE
Proposal: Add new options to dynamically change Name ID and attributes depending on SP

### ISSUE THREAD

**Issue by hamaron** [saml-idp/saml_idp#87]
# Problem
Under the current implementation of this library, the value of Name ID in a SAML response is retrieved by a lambda function defined in `/config/initializers/saml_idp.rb`, and the values of asserted attributes in the SAML response are given by lambda functions defined in either the configuration file or `principal.asserted_attributes`.

The problem I'm facing now is that it's not always the case that each SP requires the same
value as Name ID.

For example, [G-Suite requires an e-mail address as Name ID](https://support.google.com/a/answer/6330801?hl=en), but [Zoom allows use arbitrary letters for Name ID](https://support.zoom.us/hc/en-us/articles/201363003-Getting-Started-with-SSO).
While I was developing my own Idp, I want the flexibility to have users choose which value to use for Name ID for service providers like Zoom. 

The same argument is applied to asserted attributes in a SAML response.
The current implementation always produces a SAML response with all attributes
defined in either the configuration file or `principal.asserted_attributes`,
but some services such as G-Suite return an authentication error if a SAML response
includes unnecessary(undefined) attributes even when the request contains a valid Name ID.

Hence, it would be nicer if we could choose which attributes to populate in a SAML response for each SP.

# Proposal
Adding 2 new options when creating a SAML response in `SamlIdpController` like so:

```
  def create
    @saml_response = encode_response(current_user, {
        name_id: {
            format: email,
            getter: -> (principal) {
              principal.email
            }
        },
        attributes: {
            GivenName: {
                getter: :first_name
            },
            SurName: {
                getter: :last_name
            }
        }
    })
    render template: "saml_idp/idp/saml_post", layout: false
  end
```

This enables your IDP to populate both Name ID and attributes on the fly, depending on SPs and your user's configurations.

Since these are just optional values, this change wouldn't break the existing behaviors.

I'll start working on this issue in my branch, as I need this feature anyway.
I'd be really happy to hear any feedback and discuss this issue.

Thanks in advance.

**Comment by hamaron** [comment]
I opened a pull request #88 for this issue.
It'd be great if someone could take a look at it.

Thanks.

**Comment by pbssubhash** [comment]
+1

**Comment by pratikpandit05** [comment]
fwiw, we are achieving this by creating a User class and flexing the identifier based on the SP and overriding the asserted_attributes method to return only the required fields for that service provider.

**Comment by Zogoo** [comment]
> fwiw, we are achieving this by creating a User class and flexing the identifier based on the SP and overriding the asserted_attributes method to return only the required fields for that service provider. An instance of the user object can then be passed to the encode_response method

Yeah, I agree that that's one way to do it, but since we are allowing configure entire IdP app with "SamlIdp.configure" it would be nice to configure it from there.

**Comment by mjobin-mdsol** [comment]
I think we can closed, since the attached PR was merged and released
